File: 올바른괄호/javascript/20221216/src/solution.js
function test() {
    console.log("correctParentheses('()()'): true = " + correctParentheses('()()'));
    console.log("correctParentheses(')()('): false = " + correctParentheses(')()('));
}

function solution(s){
   test();
    const answer = correctParentheses(s);

    return answer;
}

function correctParentheses(string) {
    const stack = [];

    for (let i = 0; i < string.length; i += 1) {
        if (string[i] === '(') {
            stack.push('(')

            continue;
        }

        if (stack.length === 0) {
            return false;
        }

        stack.pop();
    }

    return stack.length ? false : true;
}

File: 올바른괄호/java/20221216/src/Solution.java
import java.util.*;
import java.util.stream.*;

class Solution {
    void test() {
        System.out.println("correctParentheses(\"()()\"): true = " + correctParentheses("()()"));
    }

    boolean solution(String s) {
//        test();
        boolean answer = correctParentheses(s);

        return answer;
    }

    boolean correctParentheses(String string) {
        List<String> stack = new ArrayList<>();

        for (int i = 0; i < string.length(); i += 1) {
            if (string.charAt(i) == '(') {
                stack.add(string.substring(i, i + 1));

                continue;
            }

            if (stack.size() == 0) {
                return false;
            }

            stack.remove(stack.size() - 1);
        }

        if (stack.size() == 0) {
            return true;
        }

        return false;
    }
}
